sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -320,12 +320,18 @@ class SQLAppStatusListener(
   private def onExecutionEnd(event: SparkListenerSQLExecutionEnd): Unit = {
     val SparkListenerSQLExecutionEnd(executionId, time) = event
     Option(liveExecutions.get(executionId)).foreach { exec =>
-      exec.metricsValues = aggregateMetrics(exec)
       exec.completionTime = Some(new Date(time))
-      exec.endEvents += 1
       update(exec)
 
-      removeStaleMetricsData(exec)
+      // Aggregating metrics can be expensive for large queries, so do it asynchronously.
+      // The end event count is also only updated asynchronously, to ensure that the metrics
+      // are not cleaned up by the update() call above.
+      kvstore.doAsync {
+        exec.endEvents += 1
```

Comment:
since this is updated asynchronously, is it possible that in onJobEnd we update endEvents at the same time?

Reply:
ah, good point. let me look at that.